File: findklengthsubstringswithnorepeatedcharacters.js
//Objective is to find the number of substrings of k-length and that has no
//repeated characters

let S = "havefunonleetcode", K = 5


//O(n) solution that uses a set to keep track of any repeated characters
//and a sliding window approach to traverse the array

if (K > S.length) {
    return 0
}

let count = 0
let set = new Set()
let left = 0
let right = 0

while (right < S.length) {
    //Get rid of any duplicate letters by moving left pointer
    while(set.has(S[right])) {
        set.delete(S[left++])
    }
    
    set.add(S[right])
    
    if(right - left + 1 == K) {
        set.delete(S[left++])
        count++
    }
    
    right++
}

return count;

File: findklengthsubstringswithnorepeatedcharacters.java
//Java Solution

class Solution {
    public int numKLenSubstrNoRepeats(String S, int K) {
        int count = 0;
        int left = 0;
        int right = 0;
        Set<Character> set = new HashSet<>();
        
        while (right < S.length()) {
            char c = S.charAt(right);
            while (set.contains(c)) {
                set.remove(S.charAt(left++));
            }
            
            set.add(c);
            
            if (right - left + 1 == K) {
                count++;
                set.remove(S.charAt(left++));
            }
            
            right++;
        }
        
        return count;
    }
}